lib/logstash/outputs/elasticsearch/common.rb

```diff
@@ -112,6 +111,28 @@ def event_action_tuple(event)
       [action, params, event]
     end
 
+    def validate_authentication
+      authn_options = 0
+      authn_options += 1 if @cloud_auth
+      authn_options += 1 if (@api_key && @api_key.value)
+      authn_options += 1 if (@user || (@password && @password.value))
+
+      if authn_options > 1
+        raise LogStash::ConfigurationError, 'Multiple authentication options are specified, please only use one of user/password, cloud_auth or api_key'
+      end
+
+      if @api_key && @api_key.value && @ssl   != true
+        raise(LogStash::ConfigurationError, "Using api_key authentication requires SSL/TLS secured communication using the `ssl => true` option")
+      end
+
+      if @cloud_auth
+        @user, @password = parse_user_password_from_cloud_auth(@cloud_auth)
+        # params is the plugin global params hash which will be passed to HttpClientBuilder.build
+        params['user'], params['password'] = @user, @password
+      end
+    end
+    private :validate_authentication
```

Comment:
Hey Colin, we need to consider a case were the plugin is not `register`-ed properly.
There's a use-case (monitoring) where LS just instantiates and tries to get the client.
Only realized lately (after cloud-id support shipped) and **[LS needed to make changes to get it to work](https://github.com/elastic/logstash/pull/11800/files#diff-dc3cf786acdc338cc714f141bc60c7e4R62-R65)**.
(changeset is ugly, relying even more on internals, but for the time being the only way to get cloud-id working)

We need to think about that as we'll eventually upgrade the plugin internally, ideally we would drop all of these plugin method invocations (if `build_client` would do all of the hosts filling based on *hosts, cloud-auth, api-key*). Alternative is to make `validate_authentication` public API (with a *NOTE: used by LS to build a client without plugin.register*).

Reply:
ouch, this is pretty messy. thanks for the heads-up. 
the problem we have here is with backports on elastic/logstash#11800 - by relying on internals like that the correct version on the plugin has to be installed where these changes have been made.

Instead of directly calling `fill_hosts_from_cloud_id`, `fill_user_password_from_cloud_auth` and `setup_hosts`  maybe we could have moved these into `build_client` and avoid leaking more internals and this PR here would not interfere?